File: apps/kotlin/src/main/kotlin/chapter03/DecisionTree.kt
package chapter03

import org.jetbrains.kotlinx.dataframe.AnyFrame
import org.jetbrains.kotlinx.dataframe.AnyRow
import org.jetbrains.kotlinx.dataframe.api.rows
import java.util.Locale

fun gini(labels: List<String>): Double {
    val total = labels.size.toDouble()
    return 1.0 -
        labels
            .groupingBy { it }
            .eachCount()
            .values
            .sumOf { (it / total) * (it / total) }
}

data class Split(
    val feature: String,
    val threshold: Double,
    val impurity: Double,
)

fun bestSplit(
    x: AnyFrame,
    t: List<String>,
): Split? {
    if (gini(t) == 0.0) return null
    var best: Split? = null
    for (feature in x.columnNames()) {
        val pairs =
            x[feature]
                .values()
                .map { (it as Number).toDouble() }
                .zip(t)
                .sortedBy { it.first }
        val values = pairs.map { it.first }
        val labels = pairs.map { it.second }
        for (i in 1 until pairs.size) {
            if (values[i] == values[i - 1]) continue
            val left = labels.subList(0, i)
            val right = labels.subList(i, labels.size)
            val impurity = (left.size * gini(left) + right.size * gini(right)) / pairs.size
            if (best == null || impurity < best.impurity) {
                best = Split(feature = feature, threshold = (values[i - 1] + values[i]) / 2, impurity = impurity)
            }
        }
    }
    return best
}

sealed interface Tree

data class Leaf(
    val label: String,
) : Tree

data class Node(
    val split: Split,
    val left: Tree,
    val right: Tree,
) : Tree

fun buildTree(
    x: AnyFrame,
    t: List<String>,
    maxDepth: Int?,
): Tree {
    val split =
        (if (maxDepth == 0) null else bestSplit(x, t)) ?: return Leaf(
            t
                .groupingBy { it }
                .eachCount()
                .maxBy { it.value }
                .key,
        )
    val goesLeft = x[split.feature].values().map { (it as Number).toDouble() <= split.threshold }
    val left = goesLeft.indices.filter { goesLeft[it] }
    val right = goesLeft.indices.filterNot { goesLeft[it] }
    val childDepth = maxDepth?.minus(1)
    return Node(
        split = split,
        left = buildTree(x[left], t.slice(left), childDepth),
        right = buildTree(x[right], t.slice(right), childDepth),
    )
}

fun predictOne(
    tree: Tree,
    row: AnyRow,
): String =
    when (tree) {
        is Leaf -> {
            tree.label
        }

        is Node -> {
            if ((row[tree.split.feature] as Number).toDouble() <= tree.split.threshold) {
                predictOne(tree.left, row)
            } else {
                predictOne(tree.right, row)
            }
        }
    }

class DecisionTree(
    private val maxDepth: Int? = null,
) {
    var tree: Tree? = null
        private set

    fun fit(
        x: AnyFrame,
        t: List<String>,
    ): DecisionTree {
        tree = buildTree(x, t, maxDepth)
        return this
    }

    fun predict(x: AnyFrame): List<String> {
        val fitted = checkNotNull(tree) { "fit で学習してから predict を呼んでください" }
        return x.rows().map { predictOne(fitted, it) }
    }
}

fun formatTree(
    tree: Tree,
    indent: String = "",
): String =
    when (tree) {
        is Leaf -> {
            "$indent${tree.label}"
        }

        is Node -> {
            val threshold = "%.4f".format(Locale.ROOT, tree.split.threshold)
            listOf(
                "$indent${tree.split.feature} <= $threshold",
                formatTree(tree.left, "$indent  "),
                "$indent${tree.split.feature} > $threshold",
                formatTree(tree.right, "$indent  "),
            ).joinToString("\n")
        }
    }

File: apps/python/lib/chapter03/decision_tree.py
from collections import Counter
from collections.abc import Sequence
from dataclasses import dataclass

import pandas as pd


@dataclass(frozen=True)
class Split:
    feature: str
    threshold: float
    impurity: float


def gini(labels: Sequence[str]) -> float:
    counts = Counter(labels)
    total = len(labels)
    return 1.0 - sum((count / total) ** 2 for count in counts.values())


def best_split(x: pd.DataFrame, t: pd.Series) -> Split | None:
    labels = t.to_list()
    if gini(labels) == 0.0:
        return None
    best: Split | None = None
    for feature in x.columns:
        pairs = sorted(zip(x[feature].to_list(), labels, strict=True))
        values = [value for value, _ in pairs]
        sorted_labels = [label for _, label in pairs]
        for i in range(1, len(pairs)):
            if values[i] == values[i - 1]:
                continue
            left, right = sorted_labels[:i], sorted_labels[i:]
            impurity = (len(left) * gini(left) + len(right) * gini(right)) / len(pairs)
            if best is None or impurity < best.impurity:
                threshold = (values[i - 1] + values[i]) / 2
                best = Split(
                    feature=str(feature), threshold=threshold, impurity=impurity
                )
    return best


@dataclass(frozen=True)
class Leaf:
    label: str


@dataclass(frozen=True)
class Node:
    split: Split
    left: "Leaf | Node"
    right: "Leaf | Node"


def build_tree(x: pd.DataFrame, t: pd.Series, max_depth: int | None) -> Leaf | Node:
    split = None if max_depth == 0 else best_split(x, t)
    if split is None:
        return Leaf(label=str(Counter(t.to_list()).most_common(1)[0][0]))
    goes_left = x[split.feature] <= split.threshold
    child_depth = None if max_depth is None else max_depth - 1
    return Node(
        split=split,
        left=build_tree(x[goes_left], t[goes_left], child_depth),
        right=build_tree(x[~goes_left], t[~goes_left], child_depth),
    )


def predict_one(tree: Leaf | Node, row: pd.Series) -> str:
    if isinstance(tree, Leaf):
        return tree.label
    if row[tree.split.feature] <= tree.split.threshold:
        return predict_one(tree.left, row)
    return predict_one(tree.right, row)


def format_tree(tree: Leaf | Node, indent: str = "") -> str:
    if isinstance(tree, Leaf):
        return f"{indent}{tree.label}"
    condition = f"{tree.split.feature} {{}} {tree.split.threshold:.4f}"
    return "\n".join(
        [
            indent + condition.format("<="),
            format_tree(tree.left, indent + "  "),
            indent + condition.format(">"),
            format_tree(tree.right, indent + "  "),
        ]
    )


class DecisionTree:
    def __init__(self, max_depth: int | None = None) -> None:
        self.max_depth = max_depth
        self.tree: Leaf | Node | None = None

    def fit(self, x: pd.DataFrame, t: pd.Series) -> "DecisionTree":
        self.tree = build_tree(x, t, self.max_depth)
        return self

    def predict(self, x: pd.DataFrame) -> list[str]:
        if self.tree is None:
            raise ValueError("fit で学習してから predict を呼んでください")
        tree = self.tree
        return [predict_one(tree, row) for _, row in x.iterrows()]
